### ISSUE TITLE
Primary Keys and Unique Keys

### ISSUE THREAD

**Issue by jonagoldman** [JosephSilber/bouncer#27]
Hi again, a database schema question this time.

I see in the migration file that a UNIQUE index is added to the relational tables ```user_roles```, ```user_abilities``` and ```role_abilities```.

I think this UNIQUE index is not needed because the tables already have a composite primary key.

Let me explain a bit more by showing the migration result for ```user_roles```:
```
PRIMARY KEY    (role_id, user_id)
UNIQUE         (role_id, user_id)
FOREIGN KEY    (role_id)
FOREIGN KEY    (user_id)
```

As you can see the PRIMARY KEY already makes role_id and user_id combination unique, so there is no need to do it again with a UNIQUE key.

**Comment by JosephSilber** [comment]
> the tables already have a composite primary key

Where from?

**Comment by jonagoldman** [comment]
It looks like for some reason the unique index creates an additional key in my database. Maybe it's because of the lack of a primary key? Not sure about that but I guess the unique index can be replaced with a composite primary key, achieving the same result. Thanks.

**Comment by JosephSilber** [comment]
What are the benefits of a composite primary key over a simple unique index?

**Comment by jonagoldman** [comment]
Well, I'm no DBA but each table should have a primary key as it is used to uniquely identify a row.
Generally, in regular tables its the `id` column and in relational tables its the combination of the ids columns.

A composite primary key is as simple as a unique index, its just a primary key with two columns instead of one.
A unique index is used for other things, like making sure an email is unique, or when for some reason a primary key is not enough to assure uniqueness.

Basically you are setting an unique index with two columns when you should be setting a primary key with two columns.

**Comment by Arcesilas** [comment]
Hi,

Some clues:
- http://java67.blogspot.fr/2012/10/difference-between-primary-vs-unique-key-table-sql.html
- http://dev.mysql.com/doc/refman/5.5/en/optimizing-primary-keys.html

The main difference is that Unique Key can reference NULL values, whereas PK cannot.
According to MySQL documentation:

> The primary key for a table represents the column or set of columns that you use in your most vital queries.

I understand that there is also a question of semantics ("columns that you use in your most vital queries").

When I installed and ran Bouncer migration, I only got a unique key, no Primary Key was created automatically. A PK is not mandatory. In the present case, I'm not sure it gives more performances, but it seams more accurate and appropriate in the context.

So, I think Unique Key could be replaced by a PK, but ont sure it changes anything. PK could also keep being used.
And, I repeat: I had no "duplicate key constraint" in my tables. If you don't tell MySQL to create a PK, it won't. And Laravel does not either.

So, if you have both Unique en Primary Keys, both on the same set of rows, then one of them (the PK) has been added at some time. But neither Laravel nor MySQL did add it. Do you have a package that could have added it ? Did not you had it yourself ? Is someone else working on your project with you ?

**Comment by jonagoldman** [comment]
@Arcesilas not sure why the primary keys are added to the schema, but it seems to be an issue with my environment and not related to this package.
The issue then diverged to why unique index is used instead of primary keys, and we both seem to agree this time that a composite primary key makes more sense.

**Comment by JosephSilber** [comment]
I guess what I'm asking is: what are the actual differences between the two.

I don't care either way.

I'm just curious what practical advantage there is to using composite primary keys over a unique index.

**Comment by jonagoldman** [comment]
The first rule of Database is: Always have a primary key.
The second rule of Database is: Always have a primary key.

The first actual difference is: Primary keys are faster that unique indexes.
The second actual difference is: Unique indexes can reference NULL values (and we don't want that)

**Comment by JosephSilber** [comment]
> Always have a primary key

Again, why? What practical advantage does it have?

> Primary keys are faster that unique indexes

That's only if you actually query using the primary key. Since you never actually query a pivot table with both foreign keys, this point is moot.

**Comment by Arcesilas** [comment]
> Always have a primary key

Who said this ? Which RFC or standard defines this ?

> Primary keys are faster that unique indexes

Oh ? Any benchmarks about it ?

> Since you never actually query the pivot table with both foreign keys

Oh ? You don't directly query the pivot table, but Eloquent, does, indirectly. And this is why MySQL **requires** foreign keys columns to be indexed... When a query joins a table, the joined table is therefore queried...

I'd like to mention : http://stackoverflow.com/questions/840162/should-each-and-every-table-have-a-primary-key
I think it makes more sens to have PK rather Unique Keys. In my conception of the world, Unique Keys should be used as additional constraints, such as for email data : an email should be unique in a table, it can then be used to uniquelly identified a user, but is not the column most queries concern. Relations usually use IDs which are more easily indexable (and have better performances over text columns).

**Comment by JosephSilber** [comment]
> You don't directly query the pivot table, but Eloquent, does

Except it doesn't.

To quote myself: you never actually query the pivot table with **both** foreign keys.

Eloquent will only query on one of those columns. It makes no sense to query using both of them. Since you're only querying on a single column, a composite index won't help you.

That said, we should probably have a regular index on a single column. I'll add that now.

While I'm at it, I'll swap out the unique index for a primary key, since it makes absolutely no difference anyhow.

**Comment by Arcesilas** [comment]
> you never actually query the pivot table with both foreign keys.

Really ?

```
select `abilities`.*, `user_abilities`.`user_id` as `pivot_user_id`, `user_abilities`.`ability_id` as `pivot_ability_id` from `abilities` inner join `user_abilities` on `abilities`.`id` = `user_abilities`.`ability_id` where `user_abilities`.`user_id` = ?
```

I see in the same query : ability_id and user_id from the user_abilities table.

**Comment by JosephSilber** [comment]
@Arcesilas simply referencing those two columns in the same query is not enough. I order to utilize a composite index, both columns have to appear in same clause.

In the query you provided, one column is used for the `join`, while the other column is used for the `where` clause. A composite index won't help you there.

Like I said, it makes no sense to ever query this table with a `where` clause on both columns, so the composite index is not helping us performance-wise in any way.
